### PR TITLE
Leverage suggested value literal transforms for more kinds of filter values

### PR DESCRIPTION
**Pull request by dearsaturn**
- Fix test environment incorrectly extracting `input` for graphql server
- Deprecate use of deprecated graphql utility function that determines which of a scalar union a node is.
- When using suggested value literal transforms, boolean-like and number-like values will also be coerced.
- Allow booleans, strings, numbers and null for filter values.

### CODE REVIEW

File: __tests__/app_provider.ts [hunk 1]
```diff
@@ -60,7 +62,7 @@ type KnexQueryResult = Array<{[attributeName: string]: any}>;
 const resolvers = {
     ...connectionResolvers,
     Query: {
-        async users(_: any, {input: inputArgs}: {input: IInputArgs}) {
+        async users(_: any, input: IInputArgs) {
```
Comment (dearsaturn) [inline __tests__/app_provider.ts:65]: This was an unnoticed issue in the tests where input was not being extracted by the resolver correctly.

File: __tests__/app_provider.ts [hunk 2]
```diff
@@ -70,10 +72,20 @@ const resolvers = {
...
+                isPetOwner: 'is_pet_owner'
             };
 
-            const nodeConnection = new ConnectionManager<IUserNode>(inputArgs, attributeMap);
+            const nodeConnection = new ConnectionManager<IUserNode>(input, attributeMap, {
+                resultOptions: {
+                    nodeTransformer: node => {
+                        return {
+                            ...node,
+                            isPetOwner: node.is_pet_owner
```
Comment (dearsaturn) [inline __tests__/app_provider.ts:84]: sql `is_pet_owner` -> gql `isPetOwner`

File: __tests__/app_provider.ts [hunk 3]
```diff
@@ -108,7 +120,10 @@ const appProvider = () => {
         resolvers
     });
     const app = new Koa();
-    server.applyMiddleware({app});
+    server.start().then(() => {
```
Comment (dearsaturn) [inline __tests__/app_provider.ts:123]: Newer versions of Apollo will require this order of operations.

File: src/graphql_schema.ts [hunk 4]
```diff
@@ -9,38 +11,35 @@ import {
...
-    return String(value);
-};
-
 /** @see https://stackoverflow.com/a/49911974 */
-const intString = new GraphQLScalarType({
-    name: 'IntString',
-    serialize: coerceIntString,
-    parseValue: coerceIntString,
+// tslint:disable-next-line: variable-name
+const FilterValue = new GraphQLScalarType({
```
Comment (dearsaturn) [inline src/graphql_schema.ts:16]: Perform no coercion/casting in the graphql layer. Stricten the available input types for filter value.

File: src/input_union_type.ts [hunk 5]
```diff
@@ -36,31 +31,102 @@ export default (typeName: string, inputTypes: GraphQLInputObjectType[], descript
...
                 return value;
             }
             throw generateInputTypeError(typeName, inputTypes);
         },
+        // tslint:disable-next-line: cyclomatic-complexity
         parseLiteral: ast => {
-            const inputType = inputTypes.reduce((acc: any, type) => {
-                const astClone = JSON.parse(JSON.stringify(ast));
-                try {
-                    return isValidLiteralValue(type, astClone).length === 0 ? type : acc;
```
Comment (dearsaturn) [inline src/input_union_type.ts:55]: `isValidLiteralValue` is deprecated in future versions of `graphql-js >= 15.0.0`. Now, we inspect the provided ObjectValue literal and determine whether it is a compound filter scalar or a filter scalar. The logic is as follows:

if object keys = in | or | not -> is a compound filter scalar
if object keys = field && operator && value -> is a filter scalar
else throw

File: src/input_union_type.ts [hunk 6]
```diff
@@ -36,31 +31,102 @@ export default (typeName: string, inputTypes: GraphQLInputObjectType[], descript
...
+             */
+            const isCompoundFilterScalar = ast.fields.reduce((acc, field) => {
+                if (acc) {
                     return acc;
                 }
-            }, undefined);
-            if (inputType) {
-                return valueFromAST(ast, inputType);
+
+                if (['or', 'and', 'not'].includes(field.name.value.toLowerCase())) {
```
Comment (dearsaturn) [inline src/input_union_type.ts:78]: _any_ of these indicate it is a compound filter scalar.

File: src/query_builder/knex_base.ts [hunk 7]
```diff
@@ -116,42 +117,35 @@ export default class KnexQueryBuilder implements IQueryBuilder<Knex> {
...
-        ) {
-            return [
-                (builder: QueryBuilder) => {
-                    builder.whereNull(this.computeFilterField(transformedFilter.field));
-                }
-            ];
-        }
+        const {field, operator, value} = this.filterTransformer(filter);
+
+        if (this.useSuggestedValueLiteralTransforms) {
```
Comment (dearsaturn) [inline src/query_builder/knex_base.ts:122]: This is enabled by default, and causes us to attempt to cast a provided string filter value to a native type, e.g.

```
'true' => true
'false' => false
'FALSE' => false
'1.1242' => 1.1242
'4' => 4
'A regular old string' => 'A regular old string'
```

File: src/coerce_string_value.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+
+// tslint:disable-next-line: cyclomatic-complexity
+export function coerceStringValue(value: string) {
+    /**
+     * Only try casting to float if there's at least one `.`
+     *
+     * This MUST come before parseInt because parseInt will succeed to
+     * parse a float but it will be lossy, e.g.
+     * parseInt('1.24242', 10) === 1
+     */
```
Comment (tuandoansn) [inline src/coerce_string_value.ts:11]: Will fail below:
` parseFloat('1.0 version') => 1 `

Possibly run isNaN(value) first then check if 'null','false','true', then parseInt/parseFloat.

**Review by tuandoansn — CHANGES_REQUESTED** [review]

File: src/coerce_string_value.ts (hunk 8 above)
Comment (dearsaturn) [reply src/coerce_string_value.ts:11]: doesn't the second part of the && expression protect from this?

Comment (dearsaturn) [reply src/coerce_string_value.ts:11]: Oh wait so you're saying it will succeed to parse that string because it _starts_ with a float?

Comment (tuandoansn) [reply src/coerce_string_value.ts:11]: yes, same with parseInt.
`parseInt('3Com',10) = 3`

Comment (dearsaturn) [reply src/coerce_string_value.ts:11]: This is horrible.

Comment (dearsaturn) [reply src/coerce_string_value.ts:15]: Updated to use `Number` constructor

**Review by tuandoansn — APPROVED** [review]
image

**Pull request merged**